### DISCUSSION TITLE
Package Dependency Management

### DISCUSSION THREAD

**Discussion by taytayallday** [bazel-contrib/rules_python#1888] [General]
Does rules_python have automatic package dependency management (similar to poetry)?
It looks like rules_python spins up a thread for each package is has to download. But if a package has dependencies on other packages, like pandas, will the pandas repository context (the one that gets generated after install_deps in the WORKSPACE) wait until it has the wheel file for numpy (also another repository context) until pandas starts to build? Or how does this work?

**Comment by taytayallday** [comment]
I guess my question is, the in whatever requirements.bzl file that gets generated by rules_python. Are these packages 
`_packages = [("pypi_armhf_3_10_click", "click==8.1.7"), ("pypi_armhf_3_10_pyzmq", "pyzmq==25.1.2"), ("pypi_armhf_3_10_croniter", "croniter==2.0.3"), ("pypi_armhf_3_10_python_dateutil", "python_dateutil==2.9.0.post0"),`  wich then in `install_deps()`, these get looped through, are they in dependency order? ie pandas will fall after numpy since numpy is a requirement of pandas

**Comment by aignas** [comment]
The dependencies are fetched lazily (when they are requested at bazel build time) and this order that you see in the file is just for initializing everything, at which point nothing is fetched.

The fetching of the wheels should not affect the runtime since we are just extracting the contents and parsing the METADATA files in the wheels. At runtime all of the wheels will be present and there should be no import errors.

**Comment by groodt** [comment]
In Python package installers (pip, poetry, uv, PDM, rules_python), order of installation is not guaranteed or required to be ordered. Indeed uv, is downloading and building them independently in order. So does rules_python.

rules_python requires a transitively resolved / locked set of dependencies, which then produces a graph of dependencies, which are then downloaded and installed lazily by rules_python into a runtime location suitable for an interpreter to import them (aka PYTHONPATH).
